src/ranged.cpp

```diff
@@ -2801,10 +2803,17 @@ void target_ui::action_switch_mode()
             range = turret->range();
         }
     } else {
-        ammo = relevant->gun_current_mode().target->ammo_data();
-        range = relevant->gun_current_mode().target->gun_range( you );
+        if( relevant->gun_current_mode().melee() ) {
+            mode = TargetMode::Reach;
```

Comment:
Changing UI mode from inside `target_ui::run()` is dangerous since it breaks a bunch of assumption in other places.
E.g. with `TargetMode::Fire` it's guaranteed that `relevant != nullptr` and `activity != nullptr`, but with `TargetMode::Reach` it's always `activity == nullptr`. Not sure if it _actually_ breaks anything in your case though.

Reply:
update: implemented the suggested changes. the ui now only closes if a switch from/to melee mode happens and relevant nullpointer checks were removed (wasn't able to cause a crash)